pkg/aws/services/vpclattice.go

```diff
@@ -169,3 +177,65 @@ func (d *defaultLattice) ListServiceNetworkServiceAssociationsAsList(ctx context
 
 	return result, nil
 }
+
+func (d *defaultLattice) FindServiceNetwork(ctx context.Context, name string, optionalAccountId string) (*ServiceNetworkInfo, error) {
+	input := vpclattice.ListServiceNetworksInput{}
+
+	for {
+		resp, err := d.ListServiceNetworksWithContext(ctx, &input)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, r := range resp.Items {
+			if aws.StringValue(r.Name) != name {
+				continue
+			}
+			acctIdMatches, err1 := accountIdMatches(optionalAccountId, *r.Arn)
+			if err1 != nil {
+				return nil, err1
+			}
+			if !acctIdMatches {
+				glog.V(6).Infoln("ServiceNetwork found but does not match account id ", name, r.Arn, optionalAccountId)
+				continue
+			}
+
+			glog.V(6).Infoln("Found ServiceNetwork ", name, r.Arn, optionalAccountId)
+
+			tagsInput := vpclattice.ListTagsForResourceInput{
+				ResourceArn: r.Arn,
+			}
+
+			tagsOutput, err2 := d.ListTagsForResourceWithContext(ctx, &tagsInput)
+			if err2 != nil {
+				return nil, err2
+			}
+
+			return &ServiceNetworkInfo{
+				SvcNetwork: r,
+				Tags:       tagsOutput.Tags,
+			}, nil
+		}
+
+		if resp.NextToken == nil {
+			break
+		}
+
+		input.NextToken = resp.NextToken
+	}
+
+	return nil, nil
```

Comment:
is nil,nil = not found? lattice store returns error on not found, we might have regression, but I havnt look how do we handle errors from store, might be ok-ish?

Reply:
Great question. There are only two places in the code that weren't tests and called the data store previously.

#### gateway_controller.go#254
```
	serviceNetworkStatus, err := r.datastore.GetServiceNetworkStatus(gw.Name, config.AccountID)
	if err = r.updateGatewayStatus(ctx, &serviceNetworkStatus, gw); err != nil {
		return err
	}
```
Here, previous logic actually ignored the error, which could have potentially caused issues. Since we weren't seeing crashes here, maybe there was some other invariant that always assured this codepath would resolve to an item in the cache. Ultimately, this value is only used for a message on a condition update.

#### service_manager.go#82
```
	sn, err := m.datastore.GetServiceNetworkStatus(snName, m.cloud.Config().AccountId)
	if err != nil {
		return err
	}
	assocReq := &CreateSnSvcAssocReq{
		ServiceIdentifier:        svcId,
		ServiceNetworkIdentifier: aws.String(sn.ID),
		Tags:                     m.cloud.DefaultTags(),
	}
```
While this code checks for and returns an error, what it cares about is obtaining the ID for the service network, not branching based on availability of the object in the cache. In the new code, if for some reason we can't find the service network, we'll return an error here and retry.

Reply:
I would add nullity check to service_manager.go and throw error, otherwise it might panic

Reply:
Good catch! Updated in the two places missing it. All calls should now be checking for a nil result.